// react-view-slider/lib/simple.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends2 = require('babel-runtime/helpers/extends');

var _extends3 = _interopRequireDefault(_extends2);

var _objectWithoutProperties2 = require('babel-runtime/helpers/objectWithoutProperties');

var _objectWithoutProperties3 = _interopRequireDefault(_objectWithoutProperties2);

var _toConsumableArray2 = require('babel-runtime/helpers/toConsumableArray');

var _toConsumableArray3 = _interopRequireDefault(_toConsumableArray2);

var _getPrototypeOf = require('babel-runtime/core-js/object/get-prototype-of');

var _getPrototypeOf2 = _interopRequireDefault(_getPrototypeOf);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

exports.createSimpleViewSlider = createSimpleViewSlider;

var _react = require('react');

var React = _interopRequireWildcard(_react);

var _inlineStylePrefixer = require('inline-style-prefixer');

var _inlineStylePrefixer2 = _interopRequireDefault(_inlineStylePrefixer);

var _index = require('./index');

var _index2 = _interopRequireDefault(_index);

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj.default = obj; return newObj; } }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var babelPluginFlowReactPropTypes_proptype_ViewProps = require('./index').babelPluginFlowReactPropTypes_proptype_ViewProps || require('prop-types').any;
/* eslint-env browser */

var babelPluginFlowReactPropTypes_proptype_Props = require('./index').babelPluginFlowReactPropTypes_proptype_Props || require('prop-types').any;

var babelPluginFlowReactPropTypes_proptype_Props = {
  children: require('prop-types').any,
  keepViewsMounted: require('prop-types').bool,
  animateHeight: require('prop-types').bool,
  transitionDuration: require('prop-types').number,
  transitionTimingFunction: require('prop-types').string,
  prefixer: typeof _inlineStylePrefixer2.default === 'function' ? require('prop-types').instanceOf(_inlineStylePrefixer2.default) : require('prop-types').any,
  fillParent: require('prop-types').bool,
  className: require('prop-types').string,
  style: require('prop-types').object,
  viewportClassName: require('prop-types').string,
  viewportStyle: require('prop-types').object,
  measureHeight: require('prop-types').func,
  rootRef: require('prop-types').func,
  viewportRef: require('prop-types').func
};
if (typeof exports !== 'undefined') Object.defineProperty(exports, 'babelPluginFlowReactPropTypes_proptype_Props', {
  value: babelPluginFlowReactPropTypes_proptype_Props,
  configurable: true
});
var babelPluginFlowReactPropTypes_proptype_State = {
  views: require('prop-types').node.isRequired,
  activeView: require('prop-types').number.isRequired
};
if (typeof exports !== 'undefined') Object.defineProperty(exports, 'babelPluginFlowReactPropTypes_proptype_State', {
  value: babelPluginFlowReactPropTypes_proptype_State,
  configurable: true
});


function defaultRenderView(_ref) {
  var index = _ref.index,
      key = _ref.key,
      style = _ref.style,
      ref = _ref.ref;

  return React.createElement(
    'div',
    { key: key, style: style, ref: ref },
    this.state.views[index]
  );
}

defaultRenderView.propTypes = babelPluginFlowReactPropTypes_proptype_ViewProps === require('prop-types').any ? {} : babelPluginFlowReactPropTypes_proptype_ViewProps;
function createSimpleViewSlider(ViewSlider) {
  var renderView = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : defaultRenderView;

  return function (_React$Component) {
    (0, _inherits3.default)(SimpleViewSlider, _React$Component);

    function SimpleViewSlider(props) {
      (0, _classCallCheck3.default)(this, SimpleViewSlider);

      var _this = (0, _possibleConstructorReturn3.default)(this, (SimpleViewSlider.__proto__ || (0, _getPrototypeOf2.default)(SimpleViewSlider)).call(this, props));

      _this.renderView = renderView.bind(_this);

      var child = React.Children.only(props.children);
      var activeView = parseInt(child.key);
      var views = [];
      views[activeView] = child;
      _this.state = {
        views: views,
        activeView: activeView
      };
      return _this;
    }

    (0, _createClass3.default)(SimpleViewSlider, [{
      key: 'componentWillReceiveProps',
      value: function componentWillReceiveProps(nextProps) {
        if (nextProps.children !== this.props.children) {
          var child = React.Children.only(nextProps.children);
          var _activeView = parseInt(child.key);
          var _views = [].concat((0, _toConsumableArray3.default)(this.state.views));
          _views[_activeView] = child;
          this.setState({
            views: _views,
            activeView: _activeView
          });
        }
      }
    }, {
      key: 'render',
      value: function render() {
        var _props = this.props,
            children = _props.children,
            props = (0, _objectWithoutProperties3.default)(_props, ['children']);
        var _state = this.state,
            activeView = _state.activeView,
            views = _state.views;

        return React.createElement(ViewSlider, (0, _extends3.default)({}, props, {
          renderView: this.renderView,
          numViews: views.length,
          activeView: activeView
        }));
      }
    }]);
    return SimpleViewSlider;
  }(React.Component);
}

exports.default = createSimpleViewSlider(_index2.default);
