// react-view-slider/lib/withTransitionContext.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends2 = require('babel-runtime/helpers/extends');

var _extends3 = _interopRequireDefault(_extends2);

var _getPrototypeOf = require('babel-runtime/core-js/object/get-prototype-of');

var _getPrototypeOf2 = _interopRequireDefault(_getPrototypeOf);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _react = require('react');

var React = _interopRequireWildcard(_react);

var _index = require('./index');

var _index2 = _interopRequireDefault(_index);

var _reactTransitionContext = require('react-transition-context');

var _reactTransitionContext2 = _interopRequireDefault(_reactTransitionContext);

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj.default = obj; return newObj; } }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var babelPluginFlowReactPropTypes_proptype_DefaultProps = require('./index').babelPluginFlowReactPropTypes_proptype_DefaultProps || require('prop-types').any;

var babelPluginFlowReactPropTypes_proptype_Props = require('./index').babelPluginFlowReactPropTypes_proptype_Props || require('prop-types').any;

var babelPluginFlowReactPropTypes_proptype_ViewProps = {
  index: require('prop-types').number.isRequired,
  key: require('prop-types').number.isRequired,
  active: require('prop-types').bool.isRequired,
  transitionState: require('prop-types').oneOf(['in', 'out', 'entering', 'leaving']).isRequired,
  style: require('prop-types').object.isRequired,
  ref: require('prop-types').func.isRequired
};
if (typeof exports !== 'undefined') Object.defineProperty(exports, 'babelPluginFlowReactPropTypes_proptype_ViewProps', {
  value: babelPluginFlowReactPropTypes_proptype_ViewProps,
  configurable: true
});

var ViewSliderWithTransitionContext = function (_React$Component) {
  (0, _inherits3.default)(ViewSliderWithTransitionContext, _React$Component);

  function ViewSliderWithTransitionContext() {
    var _ref;

    var _temp, _this, _ret;

    (0, _classCallCheck3.default)(this, ViewSliderWithTransitionContext);

    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    return _ret = (_temp = (_this = (0, _possibleConstructorReturn3.default)(this, (_ref = ViewSliderWithTransitionContext.__proto__ || (0, _getPrototypeOf2.default)(ViewSliderWithTransitionContext)).call.apply(_ref, [this].concat(args))), _this), _this.renderView = function (props) {
      return React.createElement(
        _reactTransitionContext2.default,
        { key: props.key, transitionState: props.transitionState },
        _this.props.renderView(props)
      );
    }, _temp), (0, _possibleConstructorReturn3.default)(_this, _ret);
  }

  (0, _createClass3.default)(ViewSliderWithTransitionContext, [{
    key: 'render',
    value: function render() {
      return React.createElement(_index2.default, (0, _extends3.default)({}, this.props, { renderView: this.renderView }));
    }
  }]);
  return ViewSliderWithTransitionContext;
}(React.Component);

ViewSliderWithTransitionContext.defaultProps = _index2.default.defaultProps;
ViewSliderWithTransitionContext.propTypes = babelPluginFlowReactPropTypes_proptype_Props;
exports.default = ViewSliderWithTransitionContext;
